File: src/main/java/com/otus/repository/CsvRepositoryImpl.java
package com.otus.repository;

import com.opencsv.CSVReader;
import com.otus.domain.QuizElement;
import lombok.RequiredArgsConstructor;

import java.io.IOException;
import java.util.List;
import java.util.stream.Collectors;

import static org.apache.commons.collections.CollectionUtils.isEmpty;

@RequiredArgsConstructor
public class CsvRepositoryImpl implements CsvRepository {

    private static final int QUESTION_INDEX = 0;
    private static final int ANSWER_INDEX = 1;
    private final CSVReader reader;
    private List<QuizElement> quizElements;

    @Override
    public List<QuizElement> getAllElements() {
        return isEmpty(quizElements) ? quizElements = readAll() : quizElements;
    }

    private List<QuizElement> readAll() {
        try {
            List<String[]> questionsAnswers = reader.readAll();
            return convertToQuizElements(questionsAnswers);
        } catch (IOException e) {
            throw new RuntimeException("Something went wrong");
        }
    }

    private List<QuizElement> convertToQuizElements(List<String[]> questionsAnswers) {
        return questionsAnswers.stream()
                .map(questionAnswer ->
                        new QuizElement(questionAnswer[QUESTION_INDEX], questionAnswer[ANSWER_INDEX]))
                .collect(Collectors.toList());
    }
}

File: src/main/java/com/otus/repository/CsvRepositoryImpl.kt
package com.otus.repository

import com.opencsv.CSVReader
import com.otus.domain.QuizElement
import com.otus.exception.InvalidDataException
import org.springframework.stereotype.Service

@Service
open class CsvRepositoryImpl(private val reader: CSVReader) : CsvRepository {

    private var quizElements: List<QuizElement> = emptyList()

    companion object {
        private const val QUESTION_INDEX: Int = 0
        private const val ANSWER_INDEX: Int = 1
    }

    override fun getAllElements(): List<QuizElement> {
        if (quizElements.isNullOrEmpty()) quizElements = readAll()

        return quizElements
    }

    private fun readAllFromFile(): List<Array<String>>? = reader.readAll() ?: emptyList()


    private fun validateData(questionsAnswers: List<Array<String>>) {
        val hasNotValidQuestionAnswer = questionsAnswers
                .any { it: Array<String> -> it.size != 2 }
                .or(questionsAnswers.isEmpty())

        if (hasNotValidQuestionAnswer) throw InvalidDataException("Data in file is not valid");
    }

    private fun readAll(): List<QuizElement> {
        return reader.readAll()
                .map { it: Array<String> -> QuizElement(it[QUESTION_INDEX], it[ANSWER_INDEX]) }
                .toList()
    }
}